### PR TITLE
[FLINK-14377] Parse Executor-relevant ProgramOptions to ConfigOptions

### PR DESCRIPTION
**Pull request by kl0u**
## What is the purpose of the change

This is part of the [FLIP-73](https://cwiki.apache.org/confluence/display/FLINK/FLIP-73%3A+Introducing+Executors+for+job+submission) effort and aims at translating the parameters currently present in the `ProgramOptions` which are relevant to job execution, into configuration options.

## Brief change log

We introduce some new `ConfigOptions` to accommodate the newly translated `ProgramOptions` to configuration parameters, and we add the `ExecutionParameterProvider` and the `ExecutionParameterProviderBuilder` which will do the actual translation.

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

**Comment by kl0u** [comment]
Please review @aljoscha and @TisonKun

File: flink-clients/src/main/java/org/apache/flink/client/cli/ExecutionParameterProviderBuilder.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+		parseClasspathURLsToConfig(commandLine, configuration);
+		parseParallelismToConfig(commandLine, configuration);
+		parseDetachedModeToConfig(commandLine, configuration);
+		parseShutdownOnExitToConfig(commandLine, configuration);
+		parseSavepointRestoreSettingsToConfig(commandLine, configuration);
+
+		return new ExecutionParameterProvider(configuration);
+	}
+
+	private static boolean parseShutdownOnExitToConfig(final CommandLine commandLine, final Configuration configuration) {
```
Comment (TisonKun) [inline flink-clients/src/main/java/org/apache/flink/client/cli/ExecutionParameterProviderBuilder.java]: Anywhere we need return values of these `parse` functions? According to `parse` prefix I would assume it is action only.

**Comment by TisonKun** [comment]
@kl0u thanks for opening this pull requests. The purpose of it looks good to me. My concern contains

1. For providing configuration, so far we use name pattern `XxxConfiguration` such as `RestConfiguration`/`JobMasterConfiguration` and so on. `ExecutionParameterProvider` looks diverge from them.

2. `fromConfiguration` is so far implemented as static methods of the `XxxConfiguration`. I wonder whether or not we have to introduce a `ExecutionParameterProviderBuilder` which doesn't follow a builder pattern as described in our [code style guide](https://lists.apache.org/x/thread.html/58e7ed148ff1df7acfacc038a5f07a3a74547caf6674c959ea6f91b4@%3Cdev.flink.apache.org%3E).

3. Since we decide to make `Configuration` the unique view of configurations in Flink, `fromCommandLine` looks like a common pattern. Shall we try to induce how to parse a `CommandLine` to `Configuration` so that it guides following development?

**Comment by kl0u** [comment]
Thanks for the review @TisonKun ! I integrated your comments.

For comment `2` and why the `Builder` is a different class, the reason is that small classes with a single purpose tend to be more maintainable and the `Builder` is already almost 200 LOC on its own.

For comment `3`, could you elaborate on what do you have in your mind? Because I am not so sure how we can have a  bulletproof scheme for parsing CLI options to ConfigOptions, that everyone has to follow, and without changing also other parts of the codebase.

**Comment by TisonKun** [comment]
Thanks for your update @kl0u !

After an investigation I actually think that the `Builder` is more like a `Factory` and here is a [patch](https://gist.github.com/TisonKun/4d574a118bb5ce00fb9ae8cd24724319) you can make use of. For a typical builder you can checkout `MiniClusterConfiguration.Builder` and so on. I'm open to learn from you opinion also :P

For comment `3`, think a bit more I would prefer defer so-call bulletproof scheme until we have more instance so that we can smoothly induce from the example instead of imagine too much right now.

**Comment by kl0u** [comment]
Thanks @TisonKun ! Let me know what you think.

**Comment by kl0u** [comment]
Thanks for the review @TisonKun !

**Comment by kl0u** [comment]
Merged.

**Pull request closed without merging**